### PR TITLE
WI #1880 Use Semantic data to optimize CrossChecker.CheckParagraph

### PR DESCRIPTION
**Pull request by mayanje**
**What have been done**

Semantic data on symbols having Paragraphs are used to optimize lookup, then the paragraph is looked up  directly in the owner paragraph domain.

These are the perfomance gains in milliseconds (dopt = optimized time, dcur = old time using previous implementation) 
mesured with the test: Parser/Programs/Cobol85/ParagraphSection.rdz.cbl 

dopt=1, dcur=91
dopt=0, dcur=1
dopt=0, dcur=2
dopt=0, dcur=0

We can notice that in  the worst case the gain is of 90%.

### CODE REVIEW

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 1]
```diff
@@ -1223,14 +1223,35 @@ public static void CheckSection(Section section)
...
+                    break;
+                default:
+                    System.Diagnostics.Debug.Assert(paragraph.SemanticData.Owner.Kind == Symbol.Kinds.Program ||
+                        paragraph.SemanticData.Owner.Kind == Symbol.Kinds.Function ||
+                        paragraph.SemanticData.Owner.Kind == Symbol.Kinds.Section);
+                    break;
+            }
+            // Look for the paragraph in the domain
+            Scopes.Container<ParagraphSymbol>.Entry entry = paragraphs.Lookup(paragraph.Name);
+            if (entry.Count > 1)
```
Comment (fm-117) [inline TypeCobol/Compiler/Diagnostics/CrossChecker.cs:1251]: Before testing the count, you can assert that the entry is not null as it contains the paragraph itself.

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 2]
```diff
@@ -1223,14 +1223,35 @@ public static void CheckSection(Section section)
...
             {
                 //Get the name of the scope to display in diagnostic message
-                var scope = string.IsNullOrEmpty(paragraph.Parent.Name) ? paragraph.Parent.ID : paragraph.Parent.Name ;
+                var scope = string.IsNullOrEmpty(paragraph.Parent.Name) ? paragraph.Parent.ID : paragraph.Parent.Name;
```
Comment (fm-117) [inline TypeCobol/Compiler/Diagnostics/CrossChecker.cs:1254]: This is the owner's name so by keeping a reference on the owner outside of the switch, you can simplify this line also.

**Review by fm-117 — CHANGES_REQUESTED** [review]

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 3]
```diff
@@ -1223,14 +1223,35 @@ public static void CheckSection(Section section)
...
+                case Symbol.Kinds.Section:
+                    {
+                        SectionSymbol section = (SectionSymbol)paragraph.SemanticData.Owner;
+                        paragraphs = section.Paragraphs;
+                    }
+                    break;
+                default:
+                    System.Diagnostics.Debug.Assert(paragraph.SemanticData.Owner.Kind == Symbol.Kinds.Program ||
+                        paragraph.SemanticData.Owner.Kind == Symbol.Kinds.Function ||
+                        paragraph.SemanticData.Owner.Kind == Symbol.Kinds.Section);
```
[the remark this answers is not in the document]
Comment (mayanje) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs]: If you like, done in c903c1d

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs (hunk 2 above)
Comment (mayanje) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs:1253]: The fact is that the diagnostic, can report 'procedure-division' as location.

**Review by fm-117 — APPROVED** [review]

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs (hunk 2 above)
Comment (fm-117) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs:1253]: Well, in fact 'procedure-division' is not a very good name, the program's name would be better.

**Review by fm-117 — CHANGES_REQUESTED** [review]

**Review by fm-117 — APPROVED** [review]
Ok to keep the current parent name mechanism for diagnostic. Sorry for the double validation.

**Comment by smedilol** [comment]
What is the test `Parser/Programs/Cobol85/ParagraphSection.rdz.cbl`? How do you run it?
What about our Performance tests ?

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 4]
```diff
@@ -1223,14 +1223,34 @@ public static void CheckSection(Section section)
...
+                        SectionSymbol section = (SectionSymbol)paragraph.SemanticData.Owner;
+                        paragraphs = section.Paragraphs;
+                    }
+                    break;
+                default:
+                    System.Diagnostics.Debug.Fail("A Program, a Function or a Section was expected as paragraph's owner.");
+                    break;
+            }
+            // Look for the paragraph in the domain
+            Scopes.Container<ParagraphSymbol>.Entry entry = paragraphs.Lookup(paragraph.Name);
```
Comment (smedilol) [inline TypeCobol/Compiler/Diagnostics/CrossChecker.cs:1248]: The only way to resolve a Symbol is the SymbolTable, we've already discuss about that.

**Review by smedilol — CHANGES_REQUESTED** [review]

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 5]
```diff
@@ -1223,14 +1223,34 @@ public static void CheckSection(Section section)
...
+            System.Diagnostics.Debug.Assert(entry != null);
+            if (entry.Count > 1)
```
[the remark this answers is not in the document]
Comment (mayanje) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs]: done in b1ae8ce

**Review by smedilol — APPROVED** [review]

**Pull request merged**